prow/test/integration/test/webhook_server_test.go

```diff
@@ -118,8 +125,142 @@ func TestWebhookServerMutateProwjob(t *testing.T) {
+		}
+	})
+
+	key := types.NamespacedName{
+		Namespace: "default",
+		Name:      "webhook-server",
+	}
+	var webhookDeployment appsv1.Deployment
+	err = kubeClient.Get(ctx, key, &webhookDeployment)
+	if err != nil {
+		t.Fatalf("could not get deployment %v", err)
+	}
+	originalProwJob := &v1.ProwJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ProwJob",
+			APIVersion: "prow.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				"foo":               "bar",
+				"default-me":        "enabled",
+				"admission-webhook": "enabled",
+			},
+			Name:      testServerAvailability,
+			Namespace: "default",
+		},
+		Spec: v1.ProwJobSpec{
+			Cluster: "foo",
+			Refs: &v1.Refs{
+				Repo: "bar",
+			},
+			Agent:  v1.TektonAgent,
+			Type:   v1.PresubmitJob,
+			Report: true,
+		},
+		Status: v1.ProwJobStatus{
+			State: v1.PendingState,
+		},
+	}
+	selector, err := labels.Parse("app = webhook-server")
+	if err != nil {
+		t.Fatal("could not parse labels")
+	}
+	webhookDeploymentCopy := webhookDeployment.DeepCopy()
+	webhookDeploymentCopy.Spec.Template.Spec.Containers[0].LivenessProbe.PeriodSeconds = 4
+	var podList corev1.PodList
+	if err := kubeClient.List(ctx, &podList, &client.ListOptions{
+		LabelSelector: selector,
+	}); err != nil {
+		t.Fatalf("could not list pods %v", err)
+	}
+	var wg sync.WaitGroup
+	wg.Add(1)
+	oldName := podList.Items[0].GetName()
+	count := 0
+	go func() {
+		defer wg.Done()
+		for {
+			exist, newPod, err := checkNewPodInCluster(oldName, kubeClient, ctx)
+			if err != nil {
+				return
+			}
+			if exist && newPod.Status.Phase == "Running" {
+				return
+			}
+			//update foo label of prowjob, a cpoy is made to ensure there is no resource version
+			originalProwJobCopy := originalProwJob.DeepCopy()
+			originalProwJobCopy.Labels["foo"] = strconv.Itoa(count)
+			if err := kubeClient.Create(ctx, originalProwJobCopy, &client.CreateOptions{}); err != nil {
+				t.Errorf("prowjob could not be created. Error: %v ", err)
+				return
+			}
+			out, err := exec.Command("kubectl", "--context="+getClusterContext(), "delete", "prowjob", testServerAvailability).CombinedOutput()
+			if err != nil {
+				t.Errorf("could not delete prowjob. Error: %v Output: %s", err, string(out))
+				return
+			}
+			count++
+		}
+	}()
+	if err := kubeClient.Patch(ctx, webhookDeploymentCopy, client.MergeFrom(&webhookDeployment), &client.PatchOptions{}); err != nil {
+		t.Fatalf("could not patch deployment")
+	}
+	wg.Wait()
+	//one more call to mutate the prowjob to check the newPod is working as expected
+	absValidBareProwJobPath, err := filepath.Abs(validBareProwJobPath)
+	if err != nil {
+		t.Fatalf("could not get absolute file path %v", err)
+	}
+	out, err := exec.Command("kubectl", "--context="+getClusterContext(), "apply", "-f", absValidBareProwJobPath).CombinedOutput()
+	if err != nil {
+		t.Fatalf("prowjob was not properly validated. Error: %v Output: %s", err, string(out))
+	}
```

Comment:
can you modify the exit condition of the for loop instead? for example instead of `return` right when a new pod is up, use a bool `isNewPod` and return at the end of the function?

Reply:
this is the section you suggested for trying another mutate

Reply:
I'm not saying not trying another mutate, please read my comment again, my point was to avoid duplicated code

Reply:
referring to this https://github.com/kubernetes/test-infra/pull/27116#:~:text=There%20is%20a,the%20newPod%20working.

Reply:
> referring to this #27116.

this links to this PR, did you mean any specific comment from this PR? If so could you provide the link to the specific comment so that I can understand the rationale better here?

Reply:
"There is a chance that the oldPod is still running when the newPod is up, so I think the end goal should be oldPod is completely deleted and the newPod is running.

Also I think we should do one more round to mutating prowjob here to make sure that it's the newPod working."